R/addDissimilarity.R

```diff
@@ -0,0 +1,312 @@
+  "addDissimilarity", signature = c("x"), function(x, method, ...)
+    standardGeneric("addDissimilarity"))
+
+#' @rdname getDissimilarity
+#' @export
+setMethod(
+    "addDissimilarity", signature = c(x = "SummarizedExperiment"),
+    function(
+          x, method, assay_name = "counts", assay.type = assay_name,
+          name = method, transposed = FALSE, tree_name = "phylo",
+          tree.name = tree_name, ...){
+    #
+    res <- getDissimilarity(
+        x, method = method, assay.type = assay.type, transposed = transposed,
+        tree.name = tree.name, ...)
+    # Add matrix to original SE
+    x <- .add_values_to_reducedDims(
+      x, values = as.matrix(res), name = name, assay.type = assay.type)
+    return(x)
+  }
+)
+
+#' @rdname getDissimilarity
+#' @export
+setGeneric(
+    "getDissimilarity", signature = c("x"), function(x, method, ...)
+        standardGeneric("getDissimilarity"))
+
+#' @rdname getDissimilarity
+#' @export
+setMethod(
+    "getDissimilarity", signature = c(x = "SummarizedExperiment"),
+    function(
+        x, method, exprs_values = "counts", assay_name = exprs_values, 
+        assay.type = assay_name, transposed = FALSE, tree_name = "phylo",
+        tree.name = tree_name, ...){
+    # Input checks
+    .check_assay_present(assay.type, x)
+    if( !.is_non_empty_string(method) ){
+        stop("'method' must be a non-empty single character value",
+            call. = FALSE)
+    }
+    if( !.is_a_bool(transposed) ){
+        stop("'na.rm' must be TRUE or FALSE.", call. = FALSE)
+    }
+    #
+    # If mrthod is unifrac, the object is TreeSE and tree was not provided by
+    # user, get tree arguments from TreeSE in addition to matrix and method.
+    if( method %in% c("unifrac") && !"tree" %in% names(list(...)) &&
+            is(x, "TreeSummarizedExperiment") ){
+        args <- .get_tree_args(
+            x,  method = method, assay.type = assay.type,
+            transposed = transposed, tree.name = tree.name, ...)
+    } else{
+      # For other methods, get only matrix and method for arguments.
+        mat <- assay(x, assay.type)
+        if( !transposed ){
+            mat <- t(mat)
+        }
+        args <- c(list(x = mat, method = method), list(...))
+    }
+    # Calculate dissimilarity based on matrix
+    mat <- do.call(getDissimilarity, args)
+    return(mat)
+    }
+)
+
+#' @rdname getDissimilarity
+#' @export
+setMethod(
+    "getDissimilarity", signature = c(x = "ANY"),
+    function(
+        x, method, exprs_values = "counts", assay_name = exprs_values, 
+        assay.type = assay_name, ...){
+    # Input check
+    if( !.is_a_string(method) ){
+        stop("'method' must be a single character value.", call. = FALSE)
+    }
+    #
+    # Calculate dissimilarity
+    mat <- .calculate_dissimilarity(mat = x, method = method, ...)
+    return(mat)
+    }
+)
+
+.calculate_dissimilarity <- function(
+        mat, method, node.label = NULL, diss.fun = NULL, tree = NULL, ...){
```

Comment:
Should we use argument name `dis.fun` instead? The fun fact is that this would match both "distance" and "dissimilarity", and some of our dissimilarity measures are distances.

Reply:
I agree